flow_ebos: remove the legacy geologic properties object

andlaus opened the pull request:
it was already almost unused (except for output). this may even imply a small performance improvement because the transmissibilities do not need to be calculated twice.

andlaus commented on opm/autodiff/BlackoilModelEbos.hpp:
```diff
@@ -1056,25 +1043,28 @@ namespace Opm {
                         const double hydrocarbon = fs.saturation(FluidSystem::oilPhaseIdx).value() + fs.saturation(FluidSystem::gasPhaseIdx).value();
-                        hcpv[region] += pv[c] * hydrocarbon;
-                        pres[region] += pv[c] * fs.pressure(FluidSystem::oilPhaseIdx).value();
+                        const double pv = intQuants.porosity().value()*ebosModel.dofTotalVolume(c);
```
note that with this approach, rock compressibility is now taken into account here.

atgeirr commented:
Need to think a little more about this, so it will not get merged today.

I think that pore volume is probably not treated correctly, it is (unfortunately) not just porosity times cell volume... Although rather than fixing this PR it might be better to work with the existing class?

blattms commented on opm/autodiff/FlowMainEbos.hpp:
```diff
@@ -546,8 +543,7 @@ namespace Opm
                 eclipse_writer_.reset(new EclipseWriter(eclState(), UgGridHelpers::createEclipseGrid( grid , inputGrid )));
-                eclipse_writer_->writeInitial(geoprops_->simProps(grid),
-                                              geoprops_->nonCartesianConnections());
+                eclipse_writer_->writeInitial(computeLegacySimProps_(), eclState().getInputNNC());
```
Note, this is not doing the same thing anymore as the grid might have added additional NNCs. These will now be gelected. `geoprops_->nonCartesianConnections()` might be a superset of `eclState().getInputNNC()`

andlaus replied:
as far as I can see, this is not the case yet? (or at least I haven't stumbled over a deck which does this yet.)

also, this is just for output, right? simulator-wise "non-geometric" NNCs are not supported by flow_ebos anyway.

finally, shouldn't this functionallity be moved to the grid?

atgeirr replied:
We should make sure that "non-geometric" NNCs *are* handled by flow_ebos. I am not sure what would be the best way to do it, though. These connections should also be dealt with in load-balancing etc., so perhaps we should make them part of the actual grid (CpGrid)?

blattms commented on opm/autodiff/FlowMainEbos.hpp:
```diff
@@ -715,6 +710,48 @@ namespace Opm
+                {"TRANY" , trany} ,
+                {"TRANZ" , tranz } };
+        }
+
```
Lots of unnecessary(?) code duplication? The differences to the existing code should be rather small. Please refactor to a common base.

andlaus replied:
good suggestion. I'd like to do this but the problem is that `eclTrans` is not available in `DerivedGeology`.

blattms commented:
Where is the [Geology::minPvFillProps_](/OPM/opm-simulators/blob/af3049b5a64928310cc069685c93d404004e59f6/opm/autodiff/GeoProps.hpp#L412-L448) done in ebos in this case?

andlaus commented:
> I think that pore volume is probably not treated correctly, it is (unfortunately) not just porosity times cell volume... Although rather than fixing this PR it might be better to work with the existing class?

ebos calculates the porosity by means of the `PORV` grid property, so this should be correct (i.e., POSITY[i] = PORV[i]/VOLUME[i]). the thing which I have more doubts about is whether the rock compressibility should be included in the FIP numbers...

andlaus commented:
> Where is the Geology::minPvFillProps_ done in ebos in this case?

here: https://github.com/OPM/ewoms/blob/master/ebos/eclproblem.hh#L1006

blattms commented:
That is not the same in a parallel run or are you working on the global grid?

blattms commented:
Sorry, I should read code before drawing conclusions. The code works on the eclipse stuff which is always global. Forget my last comment.

andlaus commented:
yeah, that code is really tricky. it also took quite a while to get this correct in the original place (`DerivedGeology`), IIRC...

andlaus commented:
jenkins run norne_ebos please

andlaus commented:
I've rebased this on top of the current master and verified that it produces exactly the same error with MPI for SPE1CASE2 as #1122.

andlaus commented:
I guess the flow_ebos reference results need to be updated because the transmissibilities are probably slightly different. I'll have to look tomorrow.

atgeirr commented:
I am not sure this is a good place to start this discussion, but I have thought a bit about the general structure of the simulators, and what is required to support interactive/script-like usage in the future. That is something we have been thinking about for sometime, but never really started on. The primary contender has been a python interface, but with things like cling it is possible to even do interactive C++.

I think that in order to support this well, the most important objects (grid, properties, model, problem etc.) must be accessible from the "user scope", in the sense that one can for example initialize a simulation from a deck, modify some properties to test something, and run the simulation with those properties.

This is fairly easily supported by the "legacy flow" way of thinking, which passes such objects typically as constructor or init() arguments to the simulator class. I think it is a bit harder in "ebos" style, whose initialization approach is more decentralized. So: if we keep the "ebos" style for initialization etc., can we still easily support a scenario as above? I guess making "getters and setters" available for most things could go some way towards it, but I don't see how we can ensure consistency then.

The reason for discussing it here, is of course that the case of "internally initialized" ebos geo props vs. the "externally initialized" legacy geo props is a good model case for the problem pattern I outlined above.

andlaus commented:
> I think that in order to support this well, the most important objects (grid, properties, model, problem etc.) must be accessible from the "user scope"

the ebos transmissibilities can already be used separately, and in fact they are actually used for grid partitioning of the `CpGrid`:

https://github.com/OPM/ewoms/blob/master/ebos/eclcpgridmanager.hh#L152

so this PR does not change much in that regard. the story is slightly different for things like porosity, but it should be very easy to factor these things out of `EclProblem` into separate classes. I'd say that we should do this "on demand".

IMO what this PR primarily changes is that it significantly reduces the amount if non-trivial redundant code, and that's a good thing, isn't it? in particular, consider the facts that the `geo_` object is not used anywhere during the "real" simulation and that the transmissibilities which it exposes are different from those which are actually used. (IIRC `DerivedGeology` missed a few fixes due to making model 2 match with the reference some months ago.)

further, you can always just create a "Simulator" object and access the quantities you're interested in via the problem.

> I guess making "getters and setters" available for most things could go some way towards it, but I don't see how we can ensure consistency then.

As I see it, instead of going through the spaghetti monster's brain of relying on "upstream" code to correctly initialize these objects before init, all code should instantiate the objects which it needs itself -- just like `EclCpGridManager` and `EclProblem` do. Maybe this sometimes leads to redundancies, but the reductions in complexity are usually very well worth it, IMO.

andlaus commented:
I've rebased this PR on top of the current master. I think that it should really get this into master in some form because without it, the geologic properties written to disk will only match those used by the simulation "by coincident" (if they match). I've verified that Norne still runs fine with this.

andlaus commented:
jenkins run this update_data please

akva2 commented:
Malformed ^

andlaus commented:
*argh* next try:

andlaus commented:
jenkins build this update_data please

andlaus commented:
weird: the norne test seems to fail because the INIT file has apparently changed. Jenkins does not seem to update the INIT file though, but I've looked at the [update script](https://github.com/OPM/opm-simulators/blob/master/tests/update_reference_data.sh#L100-L103) and it mentions the correct INIT file?

also, some reference solutions for `flow_legacy` changed, but nothing which is used by `flow_legacy` should be touched by this patch.

akva2 commented:
please do not touch anything so i can reproduce locally.

akva2 commented:
right. up till now the flow_legacy and flow_ebos produced init files have not differed, so it's a symlink in opm-data. thus cp copies to the symlink target. as flow_legacy is processed after flow_ebos, we overwrite with the file from legacy afterwards, so no change shows up. we will have to undo the symlinking.

the other changes are due to those files having been generated locally rather on jenkins last time they were pushed. they will slightly differ, even though tests changes. and the update script will update all files.

andlaus commented:
> so it's a symlink in opm-data

ah, that explains my confusion.

> the other changes are due to those files having been generated locally rather on jenkins last time they were pushed. they will slightly differ, even though tests pass. and the update script will update all files.

numerical noise!? no worries, then...

andlaus commented:
jenkins build this update_data please

andlaus commented:
jenkins build this update_data please

andlaus commented:
strange: for this PR all tests now pass locally on my machine even without OPM/opm-data#212, but Jenkins still likes to fall on its nose. 

any idea what's going on @akva2?

andlaus commented:
jenkins build this opm-data=213 please

andlaus commented:
jenkins build this opm-data=213 please

andlaus commented:
jenkins build this update_data please

andlaus commented:
@akva2: I have the suspicion that something goes belly up on the Jenkins rig with the Norne INIT file because on my machine it works/worked even without updating the reference data and update_data did not help the last time either. also, travis seems pass if it does not abort due to the usual timeout. can you investigate?

akva2 commented:
probably not today. i have mailed you the full output with logs etc, i will look into it asap if you cannot figure it out from those files.

andlaus commented:
no worries: that PR should probably get its first teeth soon, so I think that a few days more or less don't matter

andlaus commented:
jenkins build this opm-data=215 please

akva2 commented:
it you look at the full set failures in the comparison, it's rather obvious this is not round-off but a real  bug in the new code. TRANX, TRANY, TRANZ and TRANNC is quite different. it's almost as-if the scale is different or something. too much data to paste, some examples
```
Comparing TRANX...
Occurrence in first file    = 0
Occurrence in second file   = 0
Grid coordinate             = (1, 1, 1)
(first value, second value) = (68.7658, 28.1183)

Deviations exceed tolerances.
The absolute deviation is 40.6475, and the tolerance limit is 0.02.
The relative deviation is 0.5911, and the tolerance limit is 1e-05.

Occurrence in first file    = 0
Occurrence in second file   = 0
Grid coordinate             = (2, 1, 1)
(first value, second value) = (53.5751, 33.4571)

Deviations exceed tolerances.
The absolute deviation is 20.1179, and the tolerance limit is 0.02.
The relative deviation is 0.375509, and the tolerance limit is 1e-05.

Comparing TRANY...
Occurrence in first file    = 0
Occurrence in second file   = 0
Grid coordinate             = (1, 1, 1)
(first value, second value) = (71.9077, 24.7383)

Deviations exceed tolerances.
The absolute deviation is 47.1694, and the tolerance limit is 0.02.
The relative deviation is 0.655972, and the tolerance limit is 1e-05.

Occurrence in first file    = 0
Occurrence in second file   = 0
Grid coordinate             = (2, 1, 1)
(first value, second value) = (63.6425, 25.6112)

Comparing TRANZ...
Occurrence in first file    = 0
Occurrence in second file   = 0
Grid coordinate             = (1, 1, 1)
(first value, second value) = (212.795, 69.6589)

Deviations exceed tolerances.
The absolute deviation is 143.136, and the tolerance limit is 0.02.
The relative deviation is 0.672648, and the tolerance limit is 1e-05.

Occurrence in first file    = 0
Occurrence in second file   = 0
Grid coordinate             = (2, 1, 1)
(first value, second value) = (215.16, 83.3253)

Comparing TRANNNC...
Occurrence in first file    = 0
Occurrence in second file   = 0
Grid coordinate             = (3, 1, 1)
(first value, second value) = (0.0953045, 0.000737486)

Deviations exceed tolerances.
The absolute deviation is 0.094567, and the tolerance limit is 0.02.
The relative deviation is 0.992262, and the tolerance limit is 1e-05.

Occurrence in first file    = 0
Occurrence in second file   = 0
Grid coordinate             = (6, 1, 1)
(first value, second value) = (0.201114, 0.000440352)
```

this is from a local run

andlaus commented:
> it's rather obvious this is not round-off but a real bug/difference in the new code.
> [...]

I agree that this seems to be the case. it makes me wonder quite a bit why the tests pass fine without updating the reference solution on my machine, though.

anyway, the transmissibilites featured by the new INIT file are the ones which are used by the simulator, so either the differences do not matter w.r.t. summary results or I screwed the compressed->cartesian conversion code. I'll check the latter soonish.

andlaus commented:
I found a stupid mistake in the conversion code of the transmissibilities: basically it was not considering that the logically Cartesian coordinate of the cell to the inside of a face can be larger than the one on the outside. While debugging this, I've made sure that the new "raw" transmissibilites of all faces match the old ones for at least 5 decimal digits.

That still does not explain why jenkins spits out an error while my workstation does not and why the NNC transmissibilities are different for jenkins. (remember: I've checked that the first five digits of the transmissibility of every face are identical.)

andlaus commented:
jenkins build this please

andlaus commented:
jenkins build this please

andlaus commented:
I found some additional stupid bugs: use of compressed instead of logically Cartesian indices. I won't vouch for the old implementation doing this right either, so some discrepancies may be explained by this.

andlaus commented:
jenkins build this opm-data=215 please

andlaus commented:
all right: jenkins is happy now. Travis fails because it does not consider OPM/opm-data#215.

relative to the master version, the differences in TRAN[XYZ] are still present, but do not seem to be large. the huge differences in the NNC transmissibilities stem from the fact that the order in which these connections are added to opm-parser's NNC object is different from the one used by flow_legacy.

dl;dr: I'm now quite confident that the raw transmissibilities used by the simulator are identical to the old ones for Norne and somewhat less confident that the INIT file output code is correct.

@atgeirr: from your perspective, what must still be done before merging this PR?

totto82 commented:
There is a difference in TRAN[XY] for cells merged with minPV. Adding a minPvFillProps_ in ecl_transmissibility in ebos makes TRAN[XY] in flow_legacy and flow_ebos match within an acceptable tolerance for Norne. TRAN[Z] is already good. 
I can make a PR for adding minPvFillProps to ecl_transmissibility in ebos. I will continue to test TRANNNCs.

totto82 commented:
jenkins build this ewoms=181 please

akva2 commented:
Ignore the two extra failures, just a timing thing. This made TRANX|Y|Z pass, only TRANNNC breaks.

totto82 commented:
Some statistics on the TRANNNC 
NUM_TRANNNC = 15281
Average value of the TRANNNC: 5.8
MAX value: 549,26

And some on the difference between flow_ebos and flow_legacy. 
MAX difference: 0,140867
Max rel diff: 0.02
Number of TRANNNC with difference : 526

Note that the ordering of the TRANNNC changes with this PR so a direct comparison as in the test will fail.

andlaus commented:
I've rebased this on the current master and cleaned it up so that no scatter/gather operation is necessary in the parallel case anymore. the init files for parallel and sequential runs before and after the last patch are bitwise identical on my machine.

andlaus commented:
TRANZ deviates from that of E-100 for quite a few cells as well. (it was basically identical to that of flow_legacy.) for flow_ebos the Z-transmissibility was zero for all cells where the difference was larger than 1%, though. is it possible that E-100 staples some deactivated cells together in the Z direction while `CpGrid` does not?

totto82 commented:
#1179 runs without failures. i.e. the TRAN[XYZ] and TRANNNC are equivalent or nearly equivalent with the ones calculated with geoProps.  (on my machine they are exactly the same) Just the order of the TRANNNC output is changed.

atgeirr commented:
> Just the order of the TRANNNC output is changed.

Is then also the order of the NNC output changed to match?

andlaus commented:
> Is then also the order of the NNC output changed to match?

while that is possible in principle, I strongly advice to stick with the plain Dune grid interface wherever possible. Making sure that the order in which `Dune::CpGrid` iterates over the intersections of elements is impossible to match with the global ordering of faces as far as I can see, i.e., one needs to use some `Dune::CpGrid` specific APIs that come with some implicit assumptions, and that's what #1179 does.

atgeirr commented:
I'm sorry but I did not understand your answer, so perhaps my question was not quite precise... I'll try again:

If the order of the NNCs are different that should be reflected in all the output related to them, in particular the ordering for the NNC1 and NNC2 data in the EGRID file must match the ordering of TRANNNC in the INIT file. My question is: is this still the case after this change?

atgeirr commented:
Actually, I saw now that you referred to #1179, confused that with something else, sorry. I guess the answer is 'no' for this PR, right?

andlaus commented:
> in particular the ordering for the NNC1 and NNC2 data in the EGRID file must match the ordering of TRANNNC in the INIT file.

Ah, now I understand. I hope that the new INIT file is equally correct as the old one: first, the EGRID file should be consistent with the `nnc_` object automatically if transmissibilites should go to that file, and second, I doubt that the ordering is relevant here because to specify the transmissibilities of the NNCs in the INIT file, the logically-Cartesien indices of the source and destination cells must be specified. that said, since this is ECL, I would not bet a house on this.

joakim-hove commented:
> I doubt that the ordering is relevant here because to specify the transmissibilities of the NNCs in the INIT file,

The NNC information can be visualized as a list of three element tuples, where each tuple consists of two integeres corresponding to the logical cartesian index of the two cells, and the transmissibility connecting these two cells:

```
NNC1     NNC2         TRANNNC
-----------------------------
10        76           0.782
24       102           1.340
36       203           0.670
-----------------------------
```
 For this hypthetical example the vectors `NNC1 = {10, 24, 36}` and `NNC2 = {75,102,203}` will be written to the EGRID file, and the vector `TRANNNC = {0.782, 1.340, 0.670}` will be written to INIT file file. Permuting the vectors is perfectly OK - as long as the exact permutations are applied to all three vectors `NNC1, NNC2` and `TRANNC`.

andlaus commented:
once more that's "fun"! with this, the EGRID and the INIT files of the simulator should still be consistent but I'm not sure about the reference EGRID file in opm-data. I'll check...

andlaus commented:
the reference EGRID file was not updated. now it's fixed.

andlaus commented:
jenkins build this opm-data=207 please

andlaus commented:
jenkins is happy with this again.

@atgeirr: mind pressing Green?

atgeirr commented:
To be 93% safe:
 - @andlaus, the EGRID and INIT files are now in sync with respect to NNC ordering?
 - @andlaus, the linked PR OPM/opm-data#207 seems a little old, but also looks like it was updated. Are the changes in it now identical to what would be produced by the current master + this PR?
 - @totto82, you have looked at and tested this, and it ok from your side?

If the answers to all of this is 'yes', we'll merge.

andlaus commented:
> @andlaus, the EGRID and INIT files are now in sync with respect to NNC ordering?

at least ResInsight loads it just fine. I have problems to look at these transmissibilities, though.

> @andlaus, the linked PR OPM/opm-data#207 seems a little old, but also looks like it was updated. Are the changes in it now identical to what would be produced by the current master + this PR?

I accidentally used that PR instead of OPM/opm-data#215. (I was confused by all these manual reference data branches.) But yes, that currently does not matter because OPM/opm-data#207 is content-wise identical to OPM/opm-data#215 at the moment. (please merge OPM/opm-data#215, though.)

andlaus commented:
jenkins build this opm-data=215 please

joakim-hove commented:
> at least ResInsight loads it just fine. I have problems to look at these transmissibilities, though.

I have not looked at the code; but the fact that it loads in Resinsight does not prove anything with regards to possible bugs in ordering.

andlaus commented:
> I have not looked at the code; but the fact that it loads in Resinsight does not prove anything with regards to possible bugs in ordering.

yes. I basically meant to say that the new reference EGRID files are not horribly screwed up.

That said, can you -- or anyone else who has an ECLIPSE license -- check that these files load with the original toolset? it would also be great if you could also check whether the NNC transmissibilities seem to be reasonable.

totto82 commented:
I can verify that (on Norne) 
1) TRAN[XYZ] are identical or the differences are below tolerance. 
2 ) The ordering of the NNCs are different but after rearranging them into the same order. The  [NNC1, NNC2, TRANNNC] are the either identical or the differences are below tolerance.

atgeirr commented:
Thanks @totto82! With that I think we can merge this. Just so I don't make a mess: we should then merge this and OPM/opm-data#215, right?

totto82 commented:
Plus OPM/ewoms#182

atgeirr commented:
jenkins run norne opm-data=215 ewoms=182 please

andlaus commented:
> Plus OPM/ewoms#182

just did that. I somehow forgot that this was still open, sorry.

atgeirr commented:
I think that made Jenkins go belly up. I assume that PR was independent? I did not see a Jenkins run for it, which is bad... I'll try Jenkins again.

atgeirr commented:
jenkins run norne opm-data=215

andlaus commented:
weird: it still says

```
fatal: Couldn't find remote ref pull/182/merge
```

let's try again.

andlaus commented:
jenkins build this opm-data=215 please

andlaus commented:
@atgeirr: you forgot the stiff upper lip when addressing jenkins ;)

atgeirr commented:
Jenkins complains about TRANNNC comparison failure (relative error almost 1...) so the opm-data PR must be suspect.

andlaus commented:
weird: with this and OPM/opm-data#215 all tests pass locally on my machine. let's ask jenkins to do the update:

andlaus commented:
jenkins build this update_data please

andlaus commented:
jenkins build this opm-data=219 please

andlaus commented:
jenkins build this opm-data=219 please

andlaus commented:
jenkins build this opm-data=219 please

andlaus commented:
and with sugar and top!

andlaus commented:
@akva2: jenkins seems like it won't start. is anything wrong with it?

akva2 commented:
shut your pie hole and it will work fine. :D trigger seen was 'and with sugar and top!'

akva2 commented:
jenkins build this opm-data=219 please

akva2 commented:
meh. wasn't that simple. the jenkins box is having proxy issues.

akva2 commented:
well such a nice friday! this was the world conspiring against us:

1) network problems on jenkins. fixed with a restart it seems. this seemed to have been a coincidence as such.

2) a bug, sort of, in the script that opens a pr. or well. rather non-communicated assumptions.
the way the github bindings work is that after having pushed a branch (which was correctly rewritten by my scripts on the re-update), you open an issue, then convert that into a pull request. notice how the notification above only mentions an issue. that's because it couldn't convert the new issue to a pull request - because a PR already existed for the branch. after the build somebody went ahead and closed the PR manually. so no PR exists at this time, and the trigger will fail.

i manually re-opened the pr. i will now trigger the build.

akva2 commented:
jenkins build this opm-data=209 please

akva2 commented:
bam! hit the button asap before something else rear its head! :)

atgeirr commented:
Thanks a lot @akva2! I must admit I am a little confused about the multiple opm-data PRs associated with this PR by now (209, 215 and 219). The jenkins run seems to indicate that 209 is *it*, can you confirm that @andlaus? If so, please press green for both!

akva2 commented:
okay. let's try again. what happens is

- every time you run update_data, an update is force-pushed to the branch. so long, all good.
- the script then opens an *issue* on opm-data.
- this issue is then converted into a PR.

however, if there already is a PR on the branch in question, the conversion fails. ie steps 1 and 2 completely, step 3 fails. this means that you get an *issue* (215, 219), but not a new *pull request* (209). 

i have put *check for existing PR before starting issue/convert-to-pr cycle* on my todo. it's just not entirely straight forward to check for the existence, as i have to loop over some json response.

andlaus commented:
@atgeirr: OPM/opm-data#209 should be fine, because jenkins force-pushed the new reference data to OPM/opm-data#209's branch if I understood @akva2 correctly.

akva2 commented:
yes.

andlaus commented:
ok, I'll close OPM/opm-data#215 then.

akva2 commented:
note you were granted/asked to merge

andlaus commented:
didn't see this, thanks @akva2. to everyone involved: thank you for participating in this epic story!
